File: src/application/queries/doc_health_handler.py
"""Query handler for document health checking.

DocHealthHandler orchestrates the doc health check flow: loads a registry,
scans registered and unregistered docs, and returns a DocHealthReport.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Protocol, runtime_checkable

from src.domain.models.doc_health import (
    DocHealthReport,
    DocRegistryEntry,
)

if TYPE_CHECKING:
    from pathlib import Path

    from src.domain.models.doc_health import DocStatus


@runtime_checkable
class DocScannerProtocol(Protocol):
    """Interface for scanning documents on disk.

    Adapters implement this to check file existence and parse
    frontmatter for last_reviewed dates.
    """

    def load_registry(self, registry_path: Path) -> tuple[DocRegistryEntry, ...]:
        """Load document registry entries from a TOML file.

        Args:
            registry_path: Path to the doc-registry.toml file.

        Returns:
            Tuple of registry entries, empty if file is missing or broken.
        """
        ...

    def scan_registered(
        self,
        entries: tuple[DocRegistryEntry, ...],
        project_dir: Path,
    ) -> tuple[DocStatus, ...]:
        """Scan registered documents for health status.

        Args:
            entries: Registry entries to scan.
            project_dir: Project root directory.

        Returns:
            Tuple of DocStatus for each registered document.
        """
        ...

    def scan_unregistered(
        self,
        docs_dir: Path,
        registered_paths: frozenset[str],
        exclude_dirs: tuple[str, ...] = ("templates", "beads_templates"),
    ) -> tuple[DocStatus, ...]:
        """Scan for markdown files not in the registry.

        Args:
            docs_dir: Directory to scan for .md files.
            registered_paths: Paths already tracked in the registry.
            exclude_dirs: Directory names to skip.

        Returns:
            Tuple of DocStatus for unregistered documents.
        """
        ...


_DEFAULT_ENTRIES = (
    DocRegistryEntry(path="docs/PRD.md"),
    DocRegistryEntry(path="docs/DDD.md"),
    DocRegistryEntry(path="docs/ARCHITECTURE.md"),
)


class DocHealthHandler:
    """Orchestrates document health checking.

    Loads a doc registry (or uses defaults), scans registered docs,
    discovers unregistered docs, and combines into a DocHealthReport.
    """

    def __init__(self, scanner: DocScannerProtocol) -> None:
        self._scanner = scanner

    def check(self, project_dir: Path) -> DocHealthReport:
        """Check the health of project documentation.

        Args:
            project_dir: The project root directory.

        Returns:
            A DocHealthReport with statuses for all checked documents.
        """
        registry_path = project_dir / ".alty" / "maintenance" / "doc-registry.toml"
        entries = self._scanner.load_registry(registry_path)

        if not entries:
            entries = _DEFAULT_ENTRIES

        registered_statuses = self._scanner.scan_registered(entries, project_dir)

        registered_paths = frozenset(e.path for e in entries)
        docs_dir = project_dir / "docs"
        unregistered_statuses = self._scanner.scan_unregistered(
            docs_dir=docs_dir,
            registered_paths=registered_paths,
            exclude_dirs=("templates", "beads_templates"),
        )

        all_statuses = registered_statuses + unregistered_statuses
        return DocHealthReport(statuses=all_statuses)

File: internal/dochealth/application/doc_health_handler.go
// Package application provides query handlers for the DocHealth bounded context.
package application

import (
	"context"
	"fmt"
	"path/filepath"

	"github.com/alto-cli/alto/internal/dochealth/domain"
)

// DocScanner is a handler-local interface for scanning project documentation.
// It is consumed by DocHealthHandler and defined here per Go convention
// (define interfaces where consumed, not where implemented).
type DocScanner interface {
	// LoadRegistry loads document registry entries from a TOML file.
	LoadRegistry(ctx context.Context, registryPath string) ([]domain.DocRegistryEntry, error)

	// ScanRegistered scans registered documents for health status.
	ScanRegistered(ctx context.Context, entries []domain.DocRegistryEntry, projectDir string) ([]domain.DocStatus, error)

	// ScanUnregistered scans for markdown files not in the registry.
	ScanUnregistered(ctx context.Context, docsDir string, registeredPaths []string, excludeDirs []string) ([]domain.DocStatus, error)
}

// defaultEntries are used when no registry file is found.
func defaultEntries() []domain.DocRegistryEntry {
	// These are well-known defaults; review_interval_days=30 is a sensible default.
	prd, _ := domain.NewDocRegistryEntry("docs/PRD.md", "", 30)
	ddd, _ := domain.NewDocRegistryEntry("docs/DDD.md", "", 30)
	arch, _ := domain.NewDocRegistryEntry("docs/ARCHITECTURE.md", "", 30)
	return []domain.DocRegistryEntry{prd, ddd, arch}
}

// DocHealthHandler orchestrates document health checking.
// It loads a doc registry (or uses defaults), scans registered docs,
// discovers unregistered docs, and combines into a DocHealthReport.
type DocHealthHandler struct {
	scanner DocScanner
}

// NewDocHealthHandler creates a new DocHealthHandler.
func NewDocHealthHandler(scanner DocScanner) *DocHealthHandler {
	return &DocHealthHandler{scanner: scanner}
}

// Handle executes the doc health query.
func (h *DocHealthHandler) Handle(ctx context.Context, projectDir string) (domain.DocHealthReport, error) {
	registryPath := filepath.Join(projectDir, ".alto", "maintenance", "doc-registry.toml")
	entries, err := h.scanner.LoadRegistry(ctx, registryPath)
	if err != nil {
		return domain.DocHealthReport{}, fmt.Errorf("load registry: %w", err)
	}

	if len(entries) == 0 {
		entries = defaultEntries()
	}

	registeredStatuses, err := h.scanner.ScanRegistered(ctx, entries, projectDir)
	if err != nil {
		return domain.DocHealthReport{}, fmt.Errorf("scan registered docs: %w", err)
	}

	registeredPaths := make([]string, len(entries))
	for i, e := range entries {
		registeredPaths[i] = e.Path()
	}

	docsDir := filepath.Join(projectDir, "docs")
	excludeDirs := []string{"templates", "beads_templates"}
	unregisteredStatuses, err := h.scanner.ScanUnregistered(ctx, docsDir, registeredPaths, excludeDirs)
	if err != nil {
		return domain.DocHealthReport{}, fmt.Errorf("scan unregistered docs: %w", err)
	}

	allStatuses := make([]domain.DocStatus, 0, len(registeredStatuses)+len(unregisteredStatuses))
	allStatuses = append(allStatuses, registeredStatuses...)
	allStatuses = append(allStatuses, unregisteredStatuses...)

	return domain.NewDocHealthReport(allStatuses), nil
}
